Jumpscale/data/bcdb/BCDBMeta.py

```diff
@@ -43,103 +62,114 @@ def _reset_runtime_metadata(self):
         # all of this can be rebuild from the serialized information of the metastor
         self._data = None
-        # self._namespace_last_id = 0
-        self.schemas_md5 = []  # just to know which ones we already know
 
     def _load(self):
-
         self._reset_runtime_metadata()
-
         serializeddata = self._bcdb.storclient.get(0)
-
         if serializeddata is None:
             self._log_debug("save, empty schema")
-            self._data = self._schema.new()
-            self._data.name = self._bcdb.name
-            # is the initialization of the db, alsways needs to be done first
-            serializeddata = j.data.serializers.jsxdata.dumps(self._data)
+            data = {"url": {}, "md5": {}}
+            serializeddata = j.data.serializers.msgpack.dumps(data)
             self._bcdb.storclient.set(serializeddata)
         else:
             self._log_debug("schemas load from db")
-            self._data = self._schema.new(serializeddata=serializeddata)
+            self._data = j.data.serializers.msgpack.loads(serializeddata)
 
         # walk over all schema's let the j.data.schema know that these schema's exist in right order
         for s in self._data.schemas:
             self._add_to_schema_factory(s)
 
     def _schemas_in_data_print(self):
-        for s in self._data.schemas:
-            mid = self._data.url_to_mid[s.url]
-            print(" - %-35s %s   mid:%-2s hasdata:%s" % (s.url, s.md5, mid, s.hasdata))
+        print(j.core.tools._data_serializer_safe(self._data))
 
     def _save(self):
         self._log_debug("save meta:%s" % self._bcdb.name)
-        serializeddata = j.data.serializers.jsxdata.dumps(self._data)
+        serializeddata = j.data.serializers.msgpack.dumps(self._data)
         self._bcdb.storclient.set(serializeddata, 0)  # we can now always set at 0 because is for sure already there
 
-    def _schema_set(self, schema):
+    def _schema_set(self, schema, bcdbmodel):
         """
         add the schema to the metadata if it was not done yet
         :param schema:
         :return: the model id
         """
+
+        # optimized for speed, will happen quite a lot, need to know when there is change
+
         if not isinstance(schema, j.data.schema.SCHEMA_CLASS):
             raise j.exceptions.Base("schema needs to be of type: j.data.schema.SCHEMA_CLASS")
 
-        if not self._schema_exists(schema._md5):
-            s_obj = self._data.schemas.new()
-            s_obj.url = schema.url
-            s_obj.text = schema.text  # + "\n"  # only 1 \n at end
-            s_obj.md5 = schema._md5
-            s_obj.hasdata = schema.hasdata
-            self._url_mid_set(schema)
+        change = False  # we only want to save is there is a change
+
+        # deal with making sure that the md5 of this schema is registered as the newest one
+        if schema.url in self._data["url"]:
+            md5s = self._data["url"][schema.url]
+            if schema._md5 in md5s:
```

Comment:
can be joined  md5s[-1] != schema._md5 directly

Reply:
?